### ISSUE TITLE
The NB does not differentiate the smooth() operator

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9626]
When compiling [Modelica.Electrical.Analog.Examples.OpAmps.Comparator](https://libraries.openmodelica.org/branches/newInst-newBackend/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Electrical.Analog.Examples.OpAmps.Comparator.err), the following error is generated:
```
Error: Internal error NBDifferentiate.differentiateBuiltinCall2Arg failed for: smooth
```
This is an easy fix: if the first argument is at least 1, you can go ahead and return the differentiated second argument with noEvent, otherwise it means the second argument is not continuously differentiable, so any index reduction or state variable change would be incorrect, hence you should generate an error.

**Comment by kabdelhak** [comment]
I added the differentiation of `smooth()` in #9708 but the problem in this model is that following algebraic loop is detected:
```
Strict Tearing Set
****************************************
### Iteration Variables:
	[DISC] (1) Boolean $SEV_1
	 slice: {};
	[ALGB] (1) Real opAmp.v_in = vIn.p.v - potentiometer.contact.v
	 slice: {};
	[DISC] (1) Boolean $SEV_0
	 slice: {};
### Residual Equations:
	[SCAL] (1) $SEV_1 = opAmp.V0 * opAmp.v_in < opAmp.Vns ($RES_EVT_65) 
	 slice: {};
	[SCAL] (1) opAmp.v_out = smooth(0, noEvent(if $SEV_0 then opAmp.Vps else if $SEV_1 then opAmp.Vns else opAmp.V0 * opAmp.v_in)) ($RES_SIM_49) 
	 slice: {};
	[SCAL] (1) $SEV_0 = opAmp.V0 * opAmp.v_in > opAmp.Vps ($RES_EVT_67) 
	 slice: {};
### Inner Equations:
```
As we can see this cannot be correct, i assume that this issue stems from the current incorrect array connection algorithm.

**Comment by casella** [comment]
> As we can see this cannot be correct, i assume that this issue stems from the current incorrect array connection algorithm.

Just to be clear, I haven't checked these equations against the original mode, but they 100% make sense to me at first sight. The op-amp has an amplification factor of omAmp.V0 unless it hits the positive or negative supply voltages, in which case the output voltage is limited by that.

So, you can see what happens by just differentiating this once more and keeping the discontinuity, or alternatively adding the event.

I think we shouldn't be scapegoating the array connection algorithm for all things going wrong 😅.

**Comment by casella** [comment]
> > As we can see this cannot be correct, i assume that this issue stems from the current incorrect array connection algorithm.
> 
> Just to be clear, I haven't checked these equations against the original model, but they 100% make sense to me at first sight. The op-amp has an amplification factor of omAmp.V0 unless it hits the positive or negative supply voltages, in which case the output voltage is limited by that.
> 
> So, you can see what happens by just differentiating this once more and keeping the discontinuity, or alternatively adding the event.
> 
> I think we shouldn't be scapegoating the array connection algorithm for all things going wrong 😅.

**Comment by kabdelhak** [comment]
So solving this equation implicitly even though it contains a `smooth(0, ...)` is expected? I don't really understand how that makes sense to be honest. If the statement cannot be differentiated, solving it implicitly would just be guessing, right?

**Comment by casella** [comment]
> So solving this equation implicitly even though it contains a `smooth(0, ...)` is expected? I don't really understand how that makes sense to be honest. If the statement cannot be differentiated, solving it implicitly would just be guessing, right?

No, why should it be so?

There is a fundamental theory of calculus that if you want to solve f(x) = 0 and f(a) > 0, f(b) <0, and f(x) _is continuous_, then at least a solution does exist.

Of course there are multiple if's there: maybe you didn't check the conditions on f(a) and f(b), or you have a multi-dimensional problem where such a theorem does not exist, maybe there are multiple solutions, and finally there is no guarantee of convergence of Newton solver in this case. But anyway, it's by no means necessary that f(x) is differentiable to be able to solve f(x) = 0.

I would say that the only serious problems may arise if you use Newton's method when the solution is at the discontinuity, particularly if you use numerical Jacobians. Let me make a simple example
```
f(x) = c + if x > 0 then 100*x else x
```
You can draw `y = f(x)` and use the graphical interpretation of Newton's method. If you use the analytic Jacobian, it easy to see that you can solve the equation just one iteration, even in the critical case of c = 0, which has the solution x = 0 exactly at the discontinuity. However, if c = 0 and you compute the Jacobian by finite differences, it is easy to see that you may get in serious trouble.

I am thinking that may be the root cause of #9131. Of course in this case the reasoning is applied to the solution of the implicit BDF formula.

This also explains why the problem is solved when using analytic Jacobians 😃 

And, it gives us some hints about how to improve computing numerical Jacobians when you have such smoothOrder = 0 functions: you should try to avoid crossing the condition boundary when going from x to x + dx. 

A simple way to do that would be to monitor Boolean value of noEvent conditions when you go from f(x) to f(x + dx); if it does change, then change the sign of dx, so that you hopefully stay away from it. This should work in most cases, because normally it is just one scalar residual that has this problem, not many of them simultaneously.

@kabdelhak, @AnHeuermann, what do you think?

**Comment by phannebohm** [comment]
This has been fixed by #9708 and #9744 some time ago. :tada:

**Comment by casella** [comment]
Thanks for checking 😃
